bluetooth: mesh: Fix composition data page 1

HaavardRei opened the pull request:
cfg_cli: Changed buffer length check in bt_mesh_comp_p1_elem_pull
to be able to pull small final elements.


access.c: Added offset to the model relation register for vendor
models to find correct model index for CDP1
Modified macros for determining if a model is a base- or extending model.


bsim: Adds bsim test to verify that CDP1 is
encoded and decoded correctly.

Fixes #63630

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

PavelVPV commented on tests/bsim/bluetooth/mesh/src/test_cdp1.c:
```diff
@@ -0,0 +1,517 @@
+	}
+}
+
+static void verify_cdp1(struct bt_mesh_comp_p1_elem p1_elem,
+			struct bt_mesh_comp_p1_model_item mod_item,
+			struct bt_mesh_comp_p1_ext_item ext_item, struct net_buf_simple p1_dev_comp)
```
It is too much to pass these structures by values:
```suggestion
static void verify_cdp1(struct bt_mesh_comp_p1_elem *p1_elem,
			struct bt_mesh_comp_p1_model_item *mod_item,
			struct bt_mesh_comp_p1_ext_item *ext_item, struct net_buf_simple *p1_dev_comp)
```

HaavardRei replied:
Done!

PavelVPV commented on tests/bsim/bluetooth/mesh/src/test_cdp1.c:
```diff
@@ -0,0 +1,517 @@
+{
+	int elem_idx = 0;
+
+	while (bt_mesh_comp_p1_elem_pull(&p1_dev_comp, &p1_elem)) {
+		ASSERT_EQUAL(p1_elems[elem_idx].nsig, p1_elem.nsig);
+		ASSERT_EQUAL(p1_elems[elem_idx].nvnd, p1_elem.nvnd);
```
The names are little bit confusing. Not clear which variables represent hardcoded values and what were retrieved through Configuration Client. I'd suggest use some prefix, like `test_` or `test_vector`.

HaavardRei replied:
Renamed in addition to splitting the verify function into smaller functions.

PavelVPV commented on subsys/bluetooth/mesh/cfg_cli.c:
```diff
@@ -2320,7 +2320,7 @@ struct bt_mesh_mod_id_vnd bt_mesh_comp_p0_elem_mod_vnd(struct bt_mesh_comp_p0_el
 						       struct bt_mesh_comp_p1_elem *elem)
 {
-	if (buf->len < 6) {
+	if (buf->len < 4) {
```
If possible, could you move this change to a separate commit?

HaavardRei replied:
Done!

alxelax commented on tests/bsim/bluetooth/mesh/tests_scripts/comp_data/cdp1_encode_decode.sh:
```diff
@@ -0,0 +1,23 @@
+# 2. The received CDP1 is compared to a hardcoded version.
+conf=prj_mesh1d1_conf
+RunTest mesh_cdp1_test \
+	cdp1_node_data_comparison
```
Just FYI
Seems parsing\composing tests are the first candidates for unit testing. No need to use bsim since no access to the media is required.

alxelax commented on tests/bsim/bluetooth/mesh/src/test_cdp1.c:
```diff
@@ -0,0 +1,517 @@
+	}
+}
+
+static void verify_cdp1(struct bt_mesh_comp_p1_elem p1_elem,
```
Could you split this function into many simple ones? A lot of nesting levels.

alxelax replied:
functions with meaningful names to understand what each loop checks.

HaavardRei replied:
Function is now split into multiple 'verify_' function specifying what is checked.

alxelax commented on tests/bsim/bluetooth/mesh/src/test_cdp1.c:
```diff
@@ -0,0 +1,517 @@
+	.dev_key = {0xaa},
+};
+
+static const struct bt_mesh_model_cb test_model1_cb = {
```
A lot of copy-paste. Could you create a parametrized macro to create all necessary entities for one model?
Smth like:
```
#define MODEL_EXT(test_id_number) \
bt_mesh_model_extend(bt_mesh_model_find(&elems[0], test_id_number), model)

#define MODEL_NO_EXT  0

#define TEST_MODEL_DECLARE(number, ...) \
static const struct bt_mesh_model_cb test_model##number##_cb = { \
	.init = model##number##_init,    \
}; \
static const struct bt_mesh_model_op model_op##number##[] = { \
	BT_MESH_MODEL_OP_END, \
}; \
static int model##number##_init(struct bt_mesh_model *model) \
{ \
        ASSERT_OK(__VA_ARGS__); \
	return 0; \
}
```

HaavardRei replied:
Made macros for declaring models. Thanks for the suggestion!

alxelax replied:
I'm pretty sure it is possible to put functions under macro to have less copy-paste code. However, this is much better than nothing.

akredalen reviewed (commented):
Just providing some general feedback as there are already quite a few pending comments.

Good job adding the model relation offset for vendor models and updating the macros to handle model extensions correctly. Thanks! 🥇
Regarding the test, I agree with @alxelax that it would have been more suitable as a unit test, even though the JIRA task initially called for a Bsim test. I apologize for not catching this earlier. Before getting into potential ways of making the test more maintainable, should we consider making it into a unit test instead?

HaavardRei commented on subsys/bluetooth/mesh/access.c:
```diff
@@ -91,17 +91,13 @@ static struct mod_relation mod_rel_list[MOD_REL_LIST_SIZE];
-	 mod_rel_list[(idx)].idx_base == (mod)->mod_idx &&   \
-	 !(mod_rel_list[(idx)].elem_ext != (mod)->elem_idx && \
-	   mod_rel_list[(idx)].idx_ext != (mod)->mod_idx))
+	 mod_rel_list[(idx)].idx_base == (mod)->mod_idx + offset)
```
Updated!

HaavardRei commented on subsys/bluetooth/mesh/access.c:
```diff
@@ -118,6 +114,20 @@ static const struct {
+{
+	const struct bt_mesh_elem *elem = bt_mesh_model_elem(mod);
+
+	if (bt_mesh_model_find(elem, mod->id)) {
```
Changed to compare model pointers in the vnd_models array instead.

HaavardRei commented on subsys/bluetooth/mesh/access.c:
```diff
@@ -91,17 +91,13 @@ static struct mod_relation mod_rel_list[MOD_REL_LIST_SIZE];
 		 (idx)++)
 
-#define IS_MOD_BASE(mod, idx) \
+#define IS_MOD_BASE(mod, idx, offset) \
```
After discussing in person with @PavelVPV , we came to the conclusion that keeping the offset would be more efficient, and fetching the offset directly from the composition where this is available. (I. e. not calling 'get_sig_offset' unless we need to get the offset by finding the element based on the model.)

HaavardRei commented on tests/bsim/bluetooth/mesh/src/test_cdp1.c:
```diff
@@ -0,0 +1,517 @@
+
+static int model_vnd1_init(struct bt_mesh_model *model)
+{
+	ASSERT_OK(bt_mesh_model_correspond(
```
Done! Discovered a new bug when having both extend and correspond in the vendor model, which I fixed in the second commit.

PavelVPV commented:
Looks very good. Just two minor requests regarding the commits:
- please elaborate a bit the change in the commit headline in https://github.com/zephyrproject-rtos/zephyr/pull/63379/commits/d5c97b818dff88a55435394ee5a3e3629cafb25c
- please start commit message with a verb (e.g. Implement/implement, or add/Add) in https://github.com/zephyrproject-rtos/zephyr/pull/63379/commits/5e06b2f8e4f383e63e82b92e1f3d804d6c8115ac

PavelVPV commented:
@HaavardRei don't forget to resolve the conflicts.

alxelax commented on tests/bsim/bluetooth/mesh/src/test_cdp1.c:
```diff
@@ -0,0 +1,430 @@
+	err = bt_mesh_provision(net_key, 0, 0, 0, cfg.addr, cfg.dev_key);
+	if (err) {
+		FAIL("Provisioning failed (err %d)", err);
+		return;
```
do not need `return` if there is `FAIL`. Here and below.

HaavardRei replied:
Removed!

alxelax reviewed (commented):
Looks much better. Need minor clean up. Please take a look at my comment.

akredalen commented on tests/bsim/bluetooth/mesh/src/test_cdp1.c:
```diff
@@ -0,0 +1,430 @@
+		ASSERT_EQUAL(test_p1_elems[elem_idx].nsig, p1_elem->nsig);
+		ASSERT_EQUAL(test_p1_elems[elem_idx].nvnd, p1_elem->nvnd);
+
+		int mod_idx, ext_mod_idx;
```
I would declare the loop variables in the for-loop(s) instead to keep their reach limited to just within the loop. Just to be safe.
`for (int mod_idx = 0; mod_idx < p1_elem->nsig; mod_idx++) {`

HaavardRei replied:
Fixed!

akredalen reviewed (commented):
Looks good! Just a couple of suggestions :)

HaavardRei committed:
bsim: Bluetooth: mesh: Implement CDP1 test

Adds bsim test to verify that composition data page 1 (CDP1) is
encoded and decoded correctly.

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

HaavardRei committed:
Bluetooth: mesh: Change cfg_cli buffer length check

Changed buffer length check in bt_mesh_comp_p1_elem_pull.
The previous threshold would result in the
method not detecting the final element when it consisted of just one
model.

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

HaavardRei committed:
Bluetooth: mesh: access: Fix model relation register

Added offset to the model relation register for vendor
models to find correct model index
for Composition Data Page 1. The previous implementation used the
'mod_idx' from the 'bt_mesh_model' struct, which led to issues in the
model relation register due to SIG and vender models having the same
model index.
Modified existing functions related to the model relation
register to take in the offset.

Modified macros for determining if a model is a base- or
extending model.

Added check in 'add_items_to_page' to check whether the model relation
is an extension.

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

HaavardRei commented on tests/bsim/bluetooth/mesh/src/test_cdp1.c:
```diff
@@ -0,0 +1,430 @@
+{
+	struct bt_mesh_comp_p1_ext_item *test_p1_ext_item;
+
+	if (elem_idx == 0 && (mod_idx + offset) == 3) {
```
Changed to switch case :)

alxelax commented on tests/bsim/bluetooth/mesh/src/test_cdp1.c:
```diff
@@ -0,0 +1,433 @@
+		test_p1_ext_item = &test_p1_ext_mod6_mod1;
+		break;
+	default:
+		return;
```
Shouldn't be `FAIL` here instead of `return`?

HaavardRei replied:
Agreed, changed to `FAIL`.
